hi-ml/testhiml/testhiml/test_runner.py

```diff
@@ -66,6 +68,70 @@ def test_parse_and_load_model(mock_runner: Runner, model_name: Optional[str], cl
             assert mock_runner.lightning_container.model_name == model_name
 
 
+_DEFAULT_CONTAINER_NUM_NODES = 1
+_CLI_NUM_NODES = 4
+
+
+class DummyContainerWithExperimentConfigOverrides(LightningContainer):
+    container_cluster: str = param.String("default-container-cluster")
+    tag: str = param.String("default-container-tag")
+    num_nodes: int = param.Integer(_DEFAULT_CONTAINER_NUM_NODES)
+
+    def update_experiment_config(self, experiment_config: ExperimentConfig) -> None:
+        # Override parameter with different name
+        experiment_config.cluster = self.container_cluster
+        # Override parameter with clashing name
+        experiment_config.tag = self.tag
+        # Override with hard-coded value
+        experiment_config.docker_shm_size = "hardcoded-docker-shm-size"
+
+
+def test_override_experiment_config_from_container(mock_runner: Runner) -> None:
+    # Arguments partly to be set in ExperimentConfig, and partly in container.
+    args = ["",
+            "--model", "testhiml." + DummyContainerWithExperimentConfigOverrides.__name__,
+            "--container_cluster", "cli-container-cluster",
+            "--cluster", "cli-cluster",
+            "--tag", "cli-tag",
+            "--docker_shm_size", "cli-docker-shm-size",
+            "--num_nodes", str(_CLI_NUM_NODES),
+            "--mount_in_azureml", "True"
+            # "--experiment_name", "cli-experiment-name",
+            # "--workspace_name", "cli-workspace-name",
+            ]
+    with patch.object(sys, "argv", new=args):
+        mock_runner.parse_and_load_model()
+    experiment_config = mock_runner.experiment_config
+    assert experiment_config is not None
+    container = mock_runner.lightning_container
+    assert container is not None
+    assert isinstance(container, DummyContainerWithExperimentConfigOverrides)
+
+    # Current ExperimentConfig parameter priority is as follows:
+    # 1. Container
+    # 2. CLI
+    # 3. ExperimentConfig defaults
```

Comment:
I think it would make sense to have CLI as the highest priority level. Anytime someone goes through the pain of typing out something on the commandline, they probably really mean it ;-) You could achieve that by calling the `update_experiment_config` in runner line 139.

Reply:
The trouble with that is that the container may use some of its parameters to override `ExperimentConfig`. For example, what should be the expected behaviour in this case?
```python
# python runner.py --model MyContainer --batch_size 42

class MyContainer(LightningContainer):
  def update_experiment_config(experiment_config):
    experiment_config.tag = f"cool experiment with batch_size={self.batch_size}"
```

Reply:
If we choose to call `update_experiment_config()` before applying the CLI overrides to the container params, I'd at least make it a `@staticmethod` to clarify that you should only use literals or read-only values. However, I think this option is very limiting (cf. example above).

In reality, we only need the model name first to instantiate the container. How about the following:
1. Parse the runner arguments (`parser1`).
2. Instantiate container based on `--model`.
3. Override container params with remaining CLI args (`parser2`).
4. Call a `container.get_experiment_config()` method that instantiates an `ExperimentConfig` with whatever user-defined defaults.
5. Override experiment config with runner arguments (`parser1`).

Reply:
@dccastro not sure I follow the solution above, happy to follow up at voice if it helps. Overall I agree with @ant0nsc. CLI must have priority. And I think the logic with which we update arguments must be straightforward and easy to read otherwise it will become a nightmare. How often do you think we will be in the situation you describe? Is it something we already require in some of our configs?